ergo/metaculus.py

```diff
@@ -192,28 +201,81 @@ def sample_community(self):
+    ) -> pd.DataFrame:
         """
         Summarize a list of questions in a dataframe
 
         :param questions: questions to summarize
+        :param extra_columns: optional list of column names as strings
         :return: pandas dataframe summarizing the questions
         """
+
+        # It's natural to pass a string to extra_columns when you just want to add
+        # one column, e.g., ergo.MetaculusQuestion.to_dataframe(qs, extra_columns="change")
+        # Because Python treats strings similarly to lists, this can lead to cryptic errors.
+        # So I'm adding this explicit error to prevent confusion.
+        if type(extra_columns) != list:
+            raise TypeError("extra_columns must be a list")
+
         show_names = any(q.name for q in questions)
         if show_names:
-            columns = ["id", "name", "title", "resolve_time"]
+            columns = ["id", *extra_columns, "name", "title", "resolve_time"]
             data = [
-                [question.id, question.name, question.title, question.resolve_time]
+                [
+                    question.id,
+                    *[question.data[key] for key in extra_columns],
+                    question.name,
+                    question.title,
+                    question.resolve_time,
+                ]
                 for question in questions
             ]
         else:
-            columns = ["id", "title", "resolve_time"]
+            columns = ["id", *extra_columns, "title", "resolve_time"]
             data = [
-                [question.id, question.title, question.resolve_time]
+                [
+                    question.id,
+                    *[question.data[key] for key in extra_columns],
+                    question.title,
+                    question.resolve_time,
+                ]
                 for question in questions
             ]
         return pd.DataFrame(data, columns=columns)
 
+    def get_i_of_community_prediction_before(self, timestamp: int) -> Optional[int]:
+        """
+        Get index of most recent community prediction that predates timestamp
+
+        :param timestamp:
+        :return: index of most recent community prediction that predates timestamp
+        """
+        j = 1
+
+        have_not_gone_far_back_enough = True
+
+        while have_not_gone_far_back_enough:
+            if len(self.prediction_timeseries) < j:
+                # This means we went past the beginning of the timeseries
+                # and need to bring j back one.
+                j = j - 1
+                break
+
+            timestamp_difference = self.prediction_timeseries[-j]["t"] - timestamp
+
+            if timestamp_difference <= 0:
+                have_not_gone_far_back_enough = False
+            else:
+                j = j + 1
+
+        # The following happens if there aren't any community predictions yet.
+        if j == 0:
+            return None
+
+        return len(self.prediction_timeseries) - j
```

Comment:
The logic here would be easier to follow if you factor out generic functionality using a function like [`find_lt`](https://docs.python.org/3/library/bisect.html#searching-sorted-lists).

Reply:
I ended up using `bisect.bisect_left` and things are a lot simpler.